pkg/eestream/decode.go

```diff
@@ -14,28 +16,59 @@ import (
 type decodedReader struct {
 	rs     map[int]io.ReadCloser
 	es     ErasureScheme
-	inbufs map[int][]byte
 	outbuf []byte
 	err    error
+	chans  map[int]chan block
+	cb     int // current block number
 }
 
 type readerError struct {
 	i   int // reader index in the map
 	err error
 }
 
+type block struct {
+	i    int    // reader index in the map
+	num  int    // block number
+	data []byte // block data
+	err  error  // error reading the block
+}
+
 // DecodeReaders takes a map of readers and an ErasureScheme returning a
 // combined Reader. The map, 'rs', must be a mapping of erasure piece numbers
 // to erasure piece streams.
 func DecodeReaders(rs map[int]io.ReadCloser, es ErasureScheme) io.ReadCloser {
 	dr := &decodedReader{
 		rs:     rs,
 		es:     es,
-		inbufs: make(map[int][]byte, len(rs)),
 		outbuf: make([]byte, 0, es.DecodedBlockSize()),
+		chans:  make(map[int]chan block, len(rs)),
 	}
+	// Kick off a goroutine for each reader. Each reads a block from the
+	// reader and adds it to a buffered channel. If an error is read
+	// (including EOF), a block with the error is added to the channel,
+	// the channel is closed and the gourtine exits.
+	// TODO: Ensure that goroutines of slow readers really exit and don't
+	// block on adding blocks to the buffered channel.
 	for i := range rs {
-		dr.inbufs[i] = make([]byte, es.EncodedBlockSize())
+		dr.chans[i] = make(chan block, 5) // TODO make this configurable
+		go func(i int, ch chan block) {
+			// close the channel when the goroutine exits
+			defer close(ch)
+			blockNum := 0
+			for {
+				// read the next block
+				data := make([]byte, es.EncodedBlockSize())
+				_, err := io.ReadFull(dr.rs[i], data)
+				// add it to the channel
+				ch <- block{i, blockNum, data, err}
```

Comment:
here you can select on the send and a receive on the ctx.Done() cancelation channel so you don't block once the context is canceled

Reply:
Now I understand what you meant here :)